homeassistant/components/freebox/router.py

```diff
@@ -0,0 +1,381 @@
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import slugify
+
+from .const import (
+    API_VERSION,
+    APP_DESC,
+    CONN_SENSORS,
+    DOMAIN,
+    SENSOR_DEVICE_CLASS,
+    SENSOR_ICON,
+    SENSOR_NAME,
+    SENSOR_UNIT,
+    SENSOR_UPDATE,
+    STORAGE_KEY,
+    STORAGE_VERSION,
+    TEMP_SENSOR_TEMPLATE,
+    TRACKER_UPDATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class FreeboxRouter:
+    """Representation of a Freebox router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry):
+        """Initialize a Freebox router."""
+        self.hass = hass
+        self._entry = entry
+        self._host = entry.data[CONF_HOST]
+        self._port = entry.data[CONF_PORT]
+
+        self._api: Freepybox = None
+        self._name = None
+        self._mac = None
+        self._sw_v = None
+
+        # Devices
+        self._devices: Dict[str, FreeboxDevice] = {}
+
+        # Sensors
+        self._temp_sensors: Dict[str, FreeboxSensor] = {}
+        self._conn_sensors: Dict[str, FreeboxSensor] = {}
+        self._attrs = {}
+
+    async def setup(self) -> None:
+        """Set up a Freebox router."""
+        freebox_dir = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+        if not os.path.exists(freebox_dir.path):
+            await self.hass.async_add_executor_job(os.makedirs, freebox_dir.path)
+
+        token_file = self.hass.config.path(
+            f"{freebox_dir.path}/{slugify(self._host)}.conf"
+        )
+
+        self._api = Freepybox(APP_DESC, token_file, API_VERSION)
+
+        await self._api.open(self._host, self._port)
+
+        # System
+        fbx_config = await self._api.system.get_config()
+        self._mac = fbx_config["mac"]
+        self._name = fbx_config["model_info"]["pretty_name"]
+        self._sw_v = fbx_config["firmware_version"]
+
+        # Devices & sensors
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now=None) -> None:
+        """Update all Freebox platforms."""
+        await self.update_sensors()
+        await self.update_devices()
+
+    async def update_devices(self) -> None:
+        """Update Freebox devices."""
+        if self._api is None:
+            return
+
+        fbx_devices: Dict[str, any] = await self._api.lan.get_hosts_list()
+
+        # Adds the Freebox itself
+        fbx_devices.append(
+            {
+                "primary_name": self.name,
+                "l2ident": {"id": self.mac},
+                "vendor_name": self.manufacturer,
+                "host_type": "router",
+                "active": True,
+                "attrs": self._attrs,
+            }
+        )
+
+        for fbx_device in fbx_devices:
+            device_name = fbx_device["primary_name"]
```

Comment:
primary_name can be empty, that's a problem i guess if we have several like this. Could we check if it's empty and populate with mac ?

Reply:
Tried to set a name in the router but it's not updated, it has been updated after HA reboot.

Reply:
> primary_name can be empty, that's a problem i guess if we have several like this. Could we check if it's empty and populate with mac ?

I don't really like setting MAC to a device name.
Is it the same on 0.104 ?

> Tried to set a name in the router but it's not updated, it has been updated after HA reboot.

I don't get this. Setting a name of the device with no name or of the router device ?
Will investigate on naming via the UI.

Reply:
> I don't get this. Setting a name of the device with no name or of the router device ?
> Will investigate on naming via the UI.

In my freebox, i have a device that has no name, so it has no name either in HA. So i've set a name to this device in the Freebox interface (Périphériques réseau). This is updated only with a reboot of HA. Because i guess that the device tracker doesn't update the name, just the state :)

Reply:
It's exactly that.

I don't think it will happen a lot.